### PR TITLE
Refactor IREECodegenAttrs to use typed array parameters

### PR DESCRIPTION
**Pull request by MacDue**
This reworks the attributes in IREECodegenAttrs so that typed [Optional]ArrayRefParameters are used instead of untyped ArrayAttrs.

Doing this removes the manual type-checking and manual conversions to/from SmallVector<T>.

This did require a small restructuring of the lowering_config attribute, so instead of having multiple nested arrays IntegerAttrs for tiling levels, it now has one array of TilingLevelAttrs. The TilingLevelAttr then contains the tile sizes and tile interchange values for a particular level.

The TilingLevelAttr has a custom assembly format that keeps its syntax as close as possible to the previous syntax, to avoid as much test churn as possible. We plan to follow up this patch with another change that extends the TilingLevelAttr to support scalable sizes (using the canonical MLIR syntax), which will need a custom parser anyway, so we don't see this as a dealbreaker.

Example configs:

```
// No tile interchange (no change in syntax)
#iree_codegen.lowering_config<tile_sizes = [[2, 64, 64, 64],
                                            [1, 1, 1, 4],
                                            [0, 0, 0, 0]]>
// With tile interchange
// Before:
#iree_codegen.lowering_config<
   tile_sizes = [[32, 64, 0], [8, 32, 0], [0, 0, 16]],
   tile_interchange = [[1, 0, 2], [], []]>

// After:
#iree_codegen.lowering_config<tile_sizes = [
   {sizes=[32, 64, 0], interchange=[1, 0, 2]}, [8, 32, 0], [0, 0, 16]]>
```

---

These changes makes the `lowering_config` easier to extend for a follow on patch that starts adding scalable tile sizes.

### CODE REVIEW

**Comment by MacDue** [comment]
cc @banach-space

File: compiler/src/iree/compiler/Codegen/Dialect/IREECodegenAttrs.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Dialect/IREECodegenAttrs.cpp:94]: I am curious why the assembly format parser/printer didnt work here?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
This is a really nice change! Mostly looks OK to me and a good cosmetic change.
This does have some downstream effect, so we might need to stage this a bit. So marking "request changes" for now, but as soon as I get signal on that, this looks good.

File: compiler/src/iree/compiler/Codegen/Dialect/IREECodegenAttrs.cpp
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/Dialect/IREECodegenAttrs.cpp:94]: Ideally you could write:
```
let assemblyFormat = [{
    (`{` `sizes` `=` `[` $sizes `]` `,` `interchange` `=` `[` $interchange^ `]` `}`)
    : (`[` $sizes `]`)?
}];
```
But unfortunately the `assemblyFormat` limits parameters to appearing once (even across groups), so this results in: `error: duplicate parameter 'sizes'`.

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Dialect/IREECodegenAttrs.cpp:94]: I cant parse the `assemblyFormat` above, but are you trying to print/parse the `sizes` and `interchange` for each level at the same time?

Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/Dialect/IREECodegenAttrs.cpp:94]: With the new structure, the tiling levels are a list of `TilingLevelAttr`s, which contains both the sizes and interchange for a particular level. This makes the sizes and interchange 1D parameters to the `TilingLevelAttr`, which means they can be typed arrays (they have to be 1D as the typed arrays only support primitive values). 

The custom format is to make updating the tests easier (though I think looks pretty nice too :))

You could use the generated printer/parser here, but that'd only allow the format to be:
```
 #iree_codegen.lowering_config<tiling_levels = [
   {sizes=[32, 64, 0]}, {sizes=[8, 32, 0]}, {sizes=[0, 0, 16]}]>
```
The custom printer allows special-casing the "no tile interchange" case as:
```
 // Pretty much the same as the old syntax:
 #iree_codegen.lowering_config<tiling_levels = [
   [32, 64, 0], [8, 32, 0], [0, 0, 16]]>
```
This makes most tests much easier to update (just `s/tile_sizes/tiling_levels`). The full `{sizes=[32, 64, 0], interchange=[1, 0, 2]}` is only used in a handful of tests as most places don't use `tile_interchange`.  

Hope that makes sense!

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Dialect/IREECodegenAttrs.cpp:94]: Thanks for the details! THat makes sense.

**Comment by dcaballe** [comment]
Sorry, I had missed this patch! Thanks for the contribution! This is something that is going in the direction I had in mind but we may have to adjust a few things. 

Not sure I understand why the interchange side is needed to support scalable vectors but, if we want to modify the interchange representation, we have to make sure we can interchange tile loops *across the different tiling tuples/level*. For example, if we have two (random) tile levels with the following sizes [[2, 4, 6], [8, 10, 12]] we should be able to encode an interchange that results in, for example, the following loop order: 2, 4, 10, 6, 12, 8. This requires a interchange list that encodes the global position for all the tile loops.

Going one step further, the short-term changes that I had in mind for the lowering config representation included:
  * Introducing an iterator kind per dimension or similar to encode if a dimension is parallel or reduction (similar to what we have in Linalg)
  * Unifying the separate parallel and reduction tuples we have today into a single tuple.
  * Introducing a global interchange list that allows interchanges across tile tuples.

So a current lowering config like this:
```
  #iree_codegen.lowering_config<tile_sizes    = [[64, 32, 0],
                                                 [8, 16, 0],
                                                 [0, 0, 4]]>
```
would look like this:
```
  #iree_codegen.lowering_config<
        dimension_kind = [parallel, parallel, reduction],
        tile_sizes = [[64, 32, 0], [8, 16, 4]]>
        interchange = [0, 1, 5, 3, 4]>
```

Would something like this make sense to you? Would it make sense to move towards that final state? I think the current interchange is actually going in the opposite direction so we may want to revisit that.

**Comment by MacDue** [comment]
The goal of this PR is just to get the `loweing_config` into a state that's easier to modify cleanly, and it's currently not changing the functionality. The `ArrayAttr`s didn't allow for easily adding custom parsing (for the scalable sizes), and generally were a bit of a hassle (with the conversions/verification). 

I don't think the changes you propose would be any harder to accommodate with this change. I can see how I would represent them (following the changes done here), and would be happy to assist on that side :smile:

**Comment by MacDue** [comment]
For now, if we just called the tiling levels `tile_sizes` (not `tiling_levels`), moving around the interchange and adding the additional `dimension_kind` list should be fairly easy with minimal additional churn.

**Comment by banach-space** [comment]
@dcaballe  A patch that adds support for "scalable" sizes on top of this:
* https://github.com/openxla/iree/pull/15033/commits/8017fb27661f2ffd0f2b532c03c69da0f161293d

> This does have some downstream effect, so we might need to stage this a bit. So marking "request changes" for now, but as soon as I get signal on that, this looks good.

Any timelines for this, @MaheshRavishankar ? Otherwise we are bit stuck downstream with our SVE fine-tuning. Not the end of the world, but upstream would be so much better :)

**Comment by dcaballe** [comment]
Thanks for the clarification. Definitely more than happy to get help on next steps! I haven't started working on this so happy to coordinate.

The patch LGTM... given that more changes are expected around lowering config, if this is causing problems downstream, I would suggest keeping the original `tile_sizes` name to minimize the pain for now. We can take care of renaming fields all at once when we are closer to the final state? No strong opinion, though :)

**Review by dcaballe — APPROVED** [review]

**Comment by MaheshRavishankar** [comment]
cc @monorimet and @yzhang93 who could give timeline here. I'd like to land this change, so please let me know if you see any blockers.

**Comment by MaheshRavishankar** [comment]
> Sorry, I had missed this patch! Thanks for the contribution! This is something that is going in the direction I had in mind but we may have to adjust a few things.
> 
> Not sure I understand why the interchange side is needed to support scalable vectors but, if we want to modify the interchange representation, we have to make sure we can interchange tile loops _across the different tiling tuples/level_. For example, if we have two (random) tile levels with the following sizes [[2, 4, 6], [8, 10, 12]] we should be able to encode an interchange that results in, for example, the following loop order: 2, 4, 10, 6, 12, 8. This requires a interchange list that encodes the global position for all the tile loops.
> 
> Going one step further, the short-term changes that I had in mind for the lowering config representation included:
> 
> * Introducing an iterator kind per dimension or similar to encode if a dimension is parallel or reduction (similar to what we have in Linalg)
> * Unifying the separate parallel and reduction tuples we have today into a single tuple.
> * Introducing a global interchange list that allows interchanges across tile tuples.
> 
> So a current lowering config like this:
> 
> ```
>   #iree_codegen.lowering_config<tile_sizes    = [[64, 32, 0],
>                                                  [8, 16, 0],
>                                                  [0, 0, 4]]>
> ```
> 
> would look like this:
> 
> ```
>   #iree_codegen.lowering_config<
>         dimension_kind = [parallel, parallel, reduction],
>         tile_sizes = [[64, 32, 0], [8, 16, 4]]>
>         interchange = [0, 1, 5, 3, 4]>
> ```
> 
> Would something like this make sense to you? Would it make sense to move towards that final state? I think the current interchange is actually going in the opposite direction so we may want to revisit that.

I think we need to really tread lightly on the `lowering_config` changes. The intent is to not be a shadow IR. At a very high level this was the intent behind these attributes  Think of each passpipeline as a "function" that transforms the input. This "function" takes some parameters in its transformations (tile_sizes, etc.), typically related to some heuristics. The `LoweringConfig` and `TranslationInfo` is meant to capture these parameters in the IR.. `TranslationInfo` captures attributes that are for the whole translation unit, and `LoweringConfig` is the attribute that are per-op.... So as a first step, I am not really sure why we need to capture the iterator types in the `LoweringConfig`... They are not heuristics of the passpipeline.

**Comment by dcaballe** [comment]
> So as a first step, I am not really sure why we need to capture the iterator types in the LoweringConfig... They are not heuristics of the passpipeline.

We are already capturing the iterator type in the current lowering config implementation by separating tiles for parallel dimensions from tiles from reduction dimensions in different tuples (e.g., vector parallel tile sizes vs vector reduction tile sizes). Introducing the iterator type explicitly will let us unify those two tuples into one which still preserving the parallel/reduction information, which will let us scale to having more tiling levels (without the split) and introduce the global interchange order that we need for cache-level tiling.

**Comment by MaheshRavishankar** [comment]
> > So as a first step, I am not really sure why we need to capture the iterator types in the LoweringConfig... They are not heuristics of the passpipeline.
> 
> We are already capturing the iterator type in the current lowering config implementation by separating tiles for parallel dimensions from tiles from reduction dimensions in different tuples (e.g., vector parallel tile sizes vs vector reduction tile sizes). Introducing the iterator type explicitly will let us unify those two tuples into one which still preserving the parallel/reduction information, which will let us scale to having more tiling levels (without the split) and introduce the global interchange order that we need for cache-level tiling.

Yeah I have never been a fan of such a separation and "codification". The tiling levels are meant to represent the tile sizes to use at some level. The fact that some levels require loops to be parallel and some only operate on reduction dimensions is purely co-incidental. We are attaching more semantics to loweringconfig than we need to IMO

**Comment by dcaballe** [comment]
> > We are already capturing the iterator type in the current lowering config implementation by separating tiles for parallel dimensions from tiles from reduction dimensions in different tuples (e.g., vector parallel tile sizes vs vector reduction tile sizes). Introducing the iterator type explicitly will let us unify those two tuples into one which still preserving the parallel/reduction information, which will let us scale to having more tiling levels (without the split) and introduce the global interchange order that we need for cache-level tiling.
> 
> Yeah I have never been a fan of such a separation and "codification". The tiling levels are meant to represent the tile sizes to use at some level. The fact that some levels require loops to be parallel and some only operate on reduction dimensions is purely co-incidental. We are attaching more semantics to loweringconfig than we need to IMO

I guess we can get rid of that. That information is in the op holding the lowering config so it should be easy to retrieve. On the CPU side, `TilingConfig` should help preserve the existing functionality without major changes. Not so sure about the GPU side...

**Comment by MacDue** [comment]
Reverted to keeping the `tile_sizes` naming for now. It's a bit of a misnomer with these changes, but proper naming can wait till after any future changes from @dcaballe.

**Review by MaheshRavishankar — APPROVED** [review]
Ok, approving to land. Please note that if this causes issues downstream we might revert this to allow for downstream users to catch up. I'll try to not do the revert if possible, but that is a lever we might use if things get very hairy downstream. So i'd request any future changes that rely on it to be on hold for a couple of weeks if possible. 

cc @monorimet @yzhang93

**Comment by MacDue** [comment]
Once ready, could someone else merge this (I don't have commit access).

**Comment by dcaballe** [comment]
> Ok, approving to land. Please note that if this causes issues downstream we might revert this to allow for downstream users to catch up. I'll try to not do the revert if possible, but that is a lever we might use if things get very hairy downstream. So i'd request any future changes that rely on it to be on hold for a couple of weeks if possible.
> 
> cc @monorimet @yzhang93

For sure we can do as much as possible to facilitate downstream integration but reverting or blocking sound PRs for weeks because of that doesn't sound too fair to me...

**Comment by banach-space** [comment]
Many thanks for landing this @dcaballe 🙏🏻 🎉 !

> Please note that if this causes issues downstream we might revert this to allow for downstream users to catch up

This puts us in a precarious position though - we don't really know what the downstream changes are, so can't do anything to adjust this patch to reduce the friction. Also, without public timelines it is hard for us to plan our work. And all our stakeholders rely on "upstream", so maintaining our changes downstream is of little value.

We (Arm) believe that IREE will allow us to demonstrate how great SVE and SME are. However, for that to happen it is vital that IREE supports scalable auto-vectorisation. This set of non-intrusive changes is the final element of the puzzle that's missing. 

> I'll try to not do the revert if possible

I would really appreciate if you didn't and instead let us know how to help you address any issues that you may face. This is very high priority for us and we are available to assist.

Thank you everyone for reviewing/contributing and thank you for IREE - these are very exciting times for us!

-Andrzej

**Comment by MaheshRavishankar** [comment]
> Many thanks for landing this @dcaballe 🙏🏻 🎉 !
> 
> > Please note that if this causes issues downstream we might revert this to allow for downstream users to catch up
> 
> This puts us in a precarious position though - we don't really know what the downstream changes are, so can't do anything to adjust this patch to reduce the friction. Also, without public timelines it is hard for us to plan our work. And all our stakeholders rely on "upstream", so maintaining our changes downstream is of little value.
> 
> We (Arm) believe that IREE will allow us to demonstrate how great SVE and SME are. However, for that to happen it is vital that IREE supports scalable auto-vectorisation. This set of non-intrusive changes is the final element of the puzzle that's missing.
> 

Yeah totally... this is a similar problem as LLVM (and I'd approach it the same way). Issues found downstream can be cause for revert, but it has to be a "real issue"... Like I said, I will try my best to make sure that we do not revert based on downstream concerns, but as a community projects concerns of all stake holders will have to be taken into account. Will definitely loop you in and will provide context. We can decide collectively on way forward. 

> > I'll try to not do the revert if possible
> 
> I would really appreciate if you didn't and instead let us know how to help you address any issues that you may face. This is very high priority for us and we are available to assist.
> 
> Thank you everyone for reviewing/contributing and thank you for IREE - these are very exciting times for us!

Thank you folks for contributing!

**Pull request merged**
